2009-02-06  Michael Tautschnig  <mt@debian.org>

	* lib/setup-storage/Parser.pm, lib/setup-storage/Commands.pm: Make sure that
		the user hasn't configured spares/missing devices with RAID-0 and only add
		--spare-devices if at least 1 is configured. Thanks Camille Barette.
		(closes: #508192)
Index: trunk/lib/setup-storage/Commands.pm
===================================================================
--- trunk.orig/lib/setup-storage/Commands.pm
+++ trunk/lib/setup-storage/Commands.pm	
@@ -232,7 +232,7 @@
       }
       &FAI::push_command(
         "yes | mdadm --create /dev/md$id --level=$level --force --run --raid-devices="
-          . scalar(@eff_devs) . " --spare-devices=" . scalar(@spares) . " "
+          . scalar(@eff_devs) . (scalar(@spares) !=0 ? " --spare-devices=" . scalar(@spares) : "") . " "
           . join(" ", @eff_devs) . " " . join(" ", @spares),
         "$pre_req", "run_udev_/dev/md$id" );
 
Index: trunk/lib/setup-storage/Parser.pm
===================================================================
--- trunk.orig/lib/setup-storage/Parser.pm
+++ trunk/lib/setup-storage/Parser.pm	
@@ -632,6 +632,8 @@
                 ($2 =~ /spare/) and $spare = 1;
                 ($2 =~ /missing/) and $missing = 1;
               }
+              (($spare == 1 || $missing == 1) && $FAI::configs{RAID}{volumes}{$vol_id}{mode} == 0)
+                and die "RAID-0 does not support spares or missing devices\n";
               if ($missing) {
                 die "Failed to resolve $dev to a unique device name\n" if (scalar(@candidates) > 1);
                 $dev = $candidates[0] if (scalar(@candidates) == 1);
